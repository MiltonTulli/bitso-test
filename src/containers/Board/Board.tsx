import * as React from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { actions } from '../../modules/world';
import { IState } from '../../modules';
import { IMatrix } from '../../interfaces';
import { GridBoard, Navbar, Status } from '../../components';

interface IBoardMapStateToProps {
  columns: number;
  rows: number;
  matrix: IMatrix;
  islands: number;
  activeCells: number;
}
interface IBoardmapDispatchToProps {
  init: actions.InitAction;
  toggleCell: actions.ActionCreator<string>;
}
interface IBoardProps extends IBoardMapStateToProps, IBoardmapDispatchToProps {}

export class Board extends React.Component<IBoardProps> {
  componentDidMount() {
    this.props.init();
  }

  toggleCellStatus = (id: string) => {
    this.props.toggleCell(id);
    this.forceUpdate();
  };

  public render() {
    const { matrix, islands, activeCells, init, columns } = this.props;
    return (
      <>
        <Navbar />
        <Status islands={islands} activeCells={activeCells} onRefresh={init} />
        <GridBoard matrix={matrix} handleToggle={this.toggleCellStatus} columns={columns} />
      </>
    );
  }
}

const mapStateToProps = (state: IState): IBoardMapStateToProps => ({
  columns: state.worldReducer.columns,
  rows: state.worldReducer.rows,
  matrix: state.worldReducer.matrix,
  islands: state.worldReducer.islands,
  activeCells: state.worldReducer.activeCells
});

const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      init: actions.init,
      toggleCell: actions.toggleCell
    },
    dispatch
  );

export default connect<IBoardMapStateToProps, IBoardmapDispatchToProps>(
  mapStateToProps,
  mapDispatchToProps
)(Board);
