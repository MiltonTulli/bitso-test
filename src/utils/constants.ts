export const MATRIX_SIZES = ['5x5', '10x5', '10x10', '15x10', '15x15', '20x20'];
export const IS_MOBILE_BREAK = 600;
