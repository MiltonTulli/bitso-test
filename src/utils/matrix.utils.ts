import { ICell, IMatrix } from '../interfaces';

// export utils -----------------
/**
 * @function createMatrix
 * @description Create a matrix (array of arrays) given cols & rows
 * @param {Number} cols - Num of columns (x)
 * @param {Number} rows - Num of rows (y)
 */
export const createMatrix = (cols: number, rows: number): ICell[][] => {
  let matrix: ICell[][] = [];
  for (let y = 0; y < rows; y++) {
    matrix[y] = [];
    for (let x = 0; x < cols; x++) {
      matrix[y][x] = {
        id: `${x}-${y}`,
        active: false
      };
    }
  }
  return matrix;
};
/**
 * @function parseStringId
 * @description Convert string with 'N-N' format and return number for both coordenates
 * @param {String} id - 'N-N' format
 * @returns {Array} with [ x, y ] format where x & y are numbers
 */
export const parseStringId = (id: string): number[] => {
  if (!/^\d+-\d+$/.test(id)) return [];
  const coordenate = id.split('-');
  const x = Number(coordenate[0]);
  const y = Number(coordenate[1]);
  return [x, y];
};

/**
 * @function getIslands
 * @description - Get matrix groups ('islands')
 * @param {ICell[]} activeCells - Active cells
 * @returns {Array<Array>} - Return an array with arrays for each group ("island") found
 */
export const getIslands = (activeCells: ICell[]): string[][] => {
  const activeIds: string[] = activeCells.map(cell => cell.id);
  let result: string[][] = [];
  activeIds.forEach(p1 => {
    const groups = result.filter(group => group.some(p2 => isFamily(p1, p2)));
    if (groups.length) {
      result = result.filter(r => !groups.some(g => g === r));
      return result.push([...groups.flat(), p1]);
    }
    result.push([p1]);
  });
  return result;
};

export const calculateCellSize = (
  rowCount: number,
  windowWidth: number,
  spacingX: number
): number => {
  const space = windowWidth - spacingX;
  return Number(space / rowCount);
};

/**
 * @function isFamily
 * @param {String} idA - id to compre
 * @param {String} idB - id to compare
 * @returns {Boolean} - boolean value containing if id 'A' is family of id 'B'
 */
export const isFamily = (idA: string, idB: string): boolean => {
  if (idA === idB) return true;
  if (!idA) return false;
  const [x, y] = parseStringId(idA);
  switch (idB) {
    case `${x}-${y - 1}`:
    case `${x}-${y + 1}`:
    case `${x - 1}-${y}`:
    case `${x + 1}-${y}`:
      return true;
    default:
      return false;
  }
};

export const idHasFamilyOnArr = (id: string, arr: string[]): boolean => {
  if (!Array.isArray(arr)) return false;
  for (let idB of arr) {
    if (isFamily(id, idB)) {
      return true;
    }
  }
  return false;
};

export const getActiveCells = (matrix: IMatrix): ICell[] => {
  let active: ICell[] = [];
  for (let row of matrix) {
    for (let cell of row) {
      if (cell.active) {
        active.push(cell);
      }
    }
  }
  return active;
};
/**
 * Returns boolean whether two points are diagonally aligned;
 * @param { String } idA
 * @param { String } idB
 */
export const isDiagonal = (idA: string, idB: string): boolean => {
  const [x, y] = parseStringId(idA);
  return [
    `${x - 1}-${+y - 1}`,
    `${x - 1}-${+y + 1}`,
    `${x + 1}-${+y + 1}`,
    `${x + 1}-${+y - 1}`
  ].some(e => idB === e);
};
/**
 * Returns the edges of a diagonal
 * @param { String } idA
 * @param { String } idB
 */
export const getAristas = (idA: string, idB: string) => {
  const [Ax, Ay] = parseStringId(idA);
  const [Bx, By] = parseStringId(idB);
  const simbolX = Bx - Ax;
  const simbolY = By - Ay;
  return [`${Ax + simbolX}-${Ay}`, `${Ax}-${Ay + simbolY}`];
};
