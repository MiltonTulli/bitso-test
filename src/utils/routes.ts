export default {
  ROOT: '/',
  BOARD: '/board'
};
