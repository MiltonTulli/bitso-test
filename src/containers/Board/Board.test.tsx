import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { Board } from './Board';
import { createMatrix } from '../../utils/matrix.utils';
import { withRouter } from '../../utils/test';

describe('<Board>', () => {
  let props;

  beforeEach(() => {
    props = {
      columns: 5,
      rows: 5,
      matrix: createMatrix(5, 5),
      islands: 3,
      activeCells: 3,
      init: jest.fn(),
      toggleCell: jest.fn()
    };
  });

  it('Should match snapshot', () => {
    const wrapper = renderer.create(withRouter(<Board {...props} />)).toJSON();
    expect(wrapper).toMatchSnapshot();
  });

  it('Should init on didmount', () => {
    shallow(<Board {...props} />);
    expect(props.init).toHaveBeenCalled();
  });

  it('Should trigger toggleCellStatus', () => {
    const stringTest = '1-2';
    const wrapper = shallow(<Board {...props} />);
    const instance = wrapper.instance();
    (instance as Board).toggleCellStatus(stringTest);
    expect(props.toggleCell).toHaveBeenCalledWith(stringTest);
  });
});
