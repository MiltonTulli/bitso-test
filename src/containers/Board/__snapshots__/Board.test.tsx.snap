// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Board> Should match snapshot 1`] = `
Array [
  <div
    className="root-0-2-1"
  >
    <a
      className="link-0-2-3"
      href="/"
      onClick={[Function]}
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M14 7l-5 5 5 5V7z"
        />
      </svg>
      <span>
        Go back
      </span>
    </a>
    <i
      className="snes-jp-logo icon-0-2-2"
    />
  </div>,
  <div
    className="root-0-2-4"
  >
    <div
      className="nes-container is-rounded container-0-2-5"
    >
      <span>
        3
         Islands
      </span>
      <span>
        3
         Active Cells
      </span>
      <button
        className="nes-btn"
        onClick={
          [MockFunction] {
            "calls": Array [
              Array [],
            ],
            "results": Array [
              Object {
                "type": "return",
                "value": undefined,
              },
            ],
          }
        }
        type="button"
      >
        Refresh
      </button>
    </div>
  </div>,
  <div
    className="root-0-2-6"
  >
    <table
      cellSpacing="0"
      className=""
    >
      <tbody>
        <tr
          data-testid="row"
        >
          <td>
            <div
              className="root-0-2-7 root-d0-0-2-10 inactive-0-2-9"
              data-testid="cell"
              onClick={[Function]}
            />
          </td>
          <td>
            <div
              className="root-0-2-7 root-d1-0-2-11 inactive-0-2-9"
              data-testid="cell"
              onClick={[Function]}
            />
          </td>
          <td>
            <div
              className="root-0-2-7 root-d2-0-2-12 inactive-0-2-9"
              data-testid="cell"
              onClick={[Function]}
            />
          </td>
          <td>
            <div
              className="root-0-2-7 root-d3-0-2-13 inactive-0-2-9"
              data-testid="cell"
              onClick={[Function]}
            />
          </td>
          <td>
            <div
              className="root-0-2-7 root-d4-0-2-14 inactive-0-2-9"
              data-testid="cell"
              onClick={[Function]}
            />
          </td>
        </tr>
        <tr
          data-testid="row"
        >
          <td>
            <div
              className="root-0-2-7 root-d5-0-2-15 inactive-0-2-9"
              data-testid="cell"
              onClick={[Function]}
            />
          </td>
          <td>
            <div
              className="root-0-2-7 root-d6-0-2-16 inactive-0-2-9"
              data-testid="cell"
              onClick={[Function]}
            />
          </td>
          <td>
            <div
              className="root-0-2-7 root-d7-0-2-17 inactive-0-2-9"
              data-testid="cell"
              onClick={[Function]}
            />
          </td>
          <td>
            <div
              className="root-0-2-7 root-d8-0-2-18 inactive-0-2-9"
              data-testid="cell"
              onClick={[Function]}
            />
          </td>
          <td>
            <div
              className="root-0-2-7 root-d9-0-2-19 inactive-0-2-9"
              data-testid="cell"
              onClick={[Function]}
            />
          </td>
        </tr>
        <tr
          data-testid="row"
        >
          <td>
            <div
              className="root-0-2-7 root-d10-0-2-20 inactive-0-2-9"
              data-testid="cell"
              onClick={[Function]}
            />
          </td>
          <td>
            <div
              className="root-0-2-7 root-d11-0-2-21 inactive-0-2-9"
              data-testid="cell"
              onClick={[Function]}
            />
          </td>
          <td>
            <div
              className="root-0-2-7 root-d12-0-2-22 inactive-0-2-9"
              data-testid="cell"
              onClick={[Function]}
            />
          </td>
          <td>
            <div
              className="root-0-2-7 root-d13-0-2-23 inactive-0-2-9"
              data-testid="cell"
              onClick={[Function]}
            />
          </td>
          <td>
            <div
              className="root-0-2-7 root-d14-0-2-24 inactive-0-2-9"
              data-testid="cell"
              onClick={[Function]}
            />
          </td>
        </tr>
        <tr
          data-testid="row"
        >
          <td>
            <div
              className="root-0-2-7 root-d15-0-2-25 inactive-0-2-9"
              data-testid="cell"
              onClick={[Function]}
            />
          </td>
          <td>
            <div
              className="root-0-2-7 root-d16-0-2-26 inactive-0-2-9"
              data-testid="cell"
              onClick={[Function]}
            />
          </td>
          <td>
            <div
              className="root-0-2-7 root-d17-0-2-27 inactive-0-2-9"
              data-testid="cell"
              onClick={[Function]}
            />
          </td>
          <td>
            <div
              className="root-0-2-7 root-d18-0-2-28 inactive-0-2-9"
              data-testid="cell"
              onClick={[Function]}
            />
          </td>
          <td>
            <div
              className="root-0-2-7 root-d19-0-2-29 inactive-0-2-9"
              data-testid="cell"
              onClick={[Function]}
            />
          </td>
        </tr>
        <tr
          data-testid="row"
        >
          <td>
            <div
              className="root-0-2-7 root-d20-0-2-30 inactive-0-2-9"
              data-testid="cell"
              onClick={[Function]}
            />
          </td>
          <td>
            <div
              className="root-0-2-7 root-d21-0-2-31 inactive-0-2-9"
              data-testid="cell"
              onClick={[Function]}
            />
          </td>
          <td>
            <div
              className="root-0-2-7 root-d22-0-2-32 inactive-0-2-9"
              data-testid="cell"
              onClick={[Function]}
            />
          </td>
          <td>
            <div
              className="root-0-2-7 root-d23-0-2-33 inactive-0-2-9"
              data-testid="cell"
              onClick={[Function]}
            />
          </td>
          <td>
            <div
              className="root-0-2-7 root-d24-0-2-34 inactive-0-2-9"
              data-testid="cell"
              onClick={[Function]}
            />
          </td>
        </tr>
      </tbody>
    </table>
  </div>,
]
`;
